File: backend/app/api/support.py
import os
import uuid
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from fastapi.responses import FileResponse
from pydantic import BaseModel
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.core.database import get_db
from app.core.security import get_current_user, is_support_staff
from app.models.user import User, UserRole
from app.models.support import SupportMessage
from app.services.notify import push

router = APIRouter(prefix="/support", tags=["support"])

UPLOAD_DIR = "/opt/simnexus/uploads"
os.makedirs(UPLOAD_DIR, exist_ok=True)

ALLOWED_IMAGE_TYPES = {"image/jpeg", "image/png", "image/gif", "image/webp", "image/bmp"}
MAX_FILE_SIZE = 20 * 1024 * 1024  # 20 MB


class MessageIn(BaseModel):
    content: str = ""
    user_id: Optional[int] = None
    attachment_url: Optional[str] = None
    attachment_name: Optional[str] = None
    attachment_type: Optional[str] = None


class MessageOut(BaseModel):
    id: int
    user_id: int
    sender_id: int
    sender_name: str
    content: str
    is_from_user: bool
    is_read: bool
    created_at: datetime
    attachment_url: Optional[str] = None
    attachment_name: Optional[str] = None
    attachment_type: Optional[str] = None

    class Config:
        from_attributes = True


class ConversationOut(BaseModel):
    user_id: int
    username: str
    last_message: str
    last_at: datetime
    unread_count: int


# ── Upload file ───────────────────────────────────────────────────────────────

@router.post("/upload")
async def upload_file(
    file: UploadFile = File(...),
    me: User = Depends(get_current_user),
):
    data = await file.read()
    if len(data) > MAX_FILE_SIZE:
        raise HTTPException(400, "文件大小超过 20MB 限制")

    ext = os.path.splitext(file.filename or "")[1].lower()
    filename = f"{uuid.uuid4().hex}{ext}"
    path = os.path.join(UPLOAD_DIR, filename)

    with open(path, "wb") as f:
        f.write(data)

    content_type = file.content_type or ""
    att_type = "image" if content_type in ALLOWED_IMAGE_TYPES else "file"
    url = f"/api/support/files/{filename}"

    return {"url": url, "name": file.filename, "type": att_type}


# ── Serve uploaded file ───────────────────────────────────────────────────────

@router.get("/files/{filename}")
def serve_file(filename: str):
    # Prevent path traversal
    if "/" in filename or ".." in filename:
        raise HTTPException(400, "非法文件名")
    path = os.path.join(UPLOAD_DIR, filename)
    if not os.path.exists(path):
        raise HTTPException(404, "文件不存在")
    return FileResponse(path)


# ── Send message ──────────────────────────────────────────────────────────────

@router.post("/messages", response_model=MessageOut)
def send_message(body: MessageIn, db: Session = Depends(get_db), me: User = Depends(get_current_user)):
    if not body.content.strip() and not body.attachment_url:
        raise HTTPException(400, "消息或附件不能同时为空")

    if is_support_staff(me):
        if not body.user_id:
            raise HTTPException(400, "客服必须指定目标用户")
        if not db.get(User, body.user_id):
            raise HTTPException(404, "用户不存在")
        msg = SupportMessage(
            user_id=body.user_id, sender_id=me.id,
            content=body.content.strip(), is_from_user=False,
            attachment_url=body.attachment_url,
            attachment_name=body.attachment_name,
            attachment_type=body.attachment_type,
        )
    else:
        msg = SupportMessage(
            user_id=me.id, sender_id=me.id,
            content=body.content.strip(), is_from_user=True,
            attachment_url=body.attachment_url,
            attachment_name=body.attachment_name,
            attachment_type=body.attachment_type,
        )

    db.add(msg)
    db.commit()
    db.refresh(msg)
    preview = body.content.strip()[:40] or (f"[{body.attachment_name}]" if body.attachment_name else "[附件]")
    if is_support_staff(me):
        push("support_reply", "客服已回复您的咨询", preview,
             audience="user", target_user_id=body.user_id)
    else:
        # Notify admins about new user message
        push("support_msg", f"用户咨询：{me.username}", preview, audience="support")
    return _to_out(msg, db)


# ── Get messages ──────────────────────────────────────────────────────────────

@router.get("/messages", response_model=list[MessageOut])
def get_messages(
    user_id: Optional[int] = None,
    since_id: Optional[int] = None,
    db: Session = Depends(get_db),
    me: User = Depends(get_current_user),
):
    if is_support_staff(me):
        if not user_id:
            raise HTTPException(400, "需要指定 user_id")
        q = db.query(SupportMessage).filter(SupportMessage.user_id == user_id)
    else:
        q = db.query(SupportMessage).filter(SupportMessage.user_id == me.id)

    if since_id:
        q = q.filter(SupportMessage.id > since_id)

    return [_to_out(m, db) for m in q.order_by(SupportMessage.created_at.asc()).all()]


# ── Mark as read ──────────────────────────────────────────────────────────────

@router.post("/messages/read")
def mark_read(user_id: Optional[int] = None, db: Session = Depends(get_db), me: User = Depends(get_current_user)):
    if is_support_staff(me):
        if not user_id:
            raise HTTPException(400, "需要 user_id")
        db.query(SupportMessage).filter(
            SupportMessage.user_id == user_id,
            SupportMessage.is_from_user == True,
            SupportMessage.is_read == False,
        ).update({"is_read": True})
    else:
        db.query(SupportMessage).filter(
            SupportMessage.user_id == me.id,
            SupportMessage.is_from_user == False,
            SupportMessage.is_read == False,
        ).update({"is_read": True})
    db.commit()
    return {"ok": True}


# ── Unread count ──────────────────────────────────────────────────────────────

@router.get("/unread")
def unread_count(db: Session = Depends(get_db), me: User = Depends(get_current_user)):
    if is_support_staff(me):
        count = db.query(func.count(SupportMessage.id)).filter(
            SupportMessage.is_from_user == True, SupportMessage.is_read == False,
        ).scalar()
    else:
        count = db.query(func.count(SupportMessage.id)).filter(
            SupportMessage.user_id == me.id,
            SupportMessage.is_from_user == False, SupportMessage.is_read == False,
        ).scalar()
    return {"count": count or 0}


# ── Conversations (admin) ─────────────────────────────────────────────────────

@router.get("/conversations", response_model=list[ConversationOut])
def list_conversations(db: Session = Depends(get_db), me: User = Depends(get_current_user)):
    if not is_support_staff(me):
        raise HTTPException(403, "无客服权限")
    from sqlalchemy import distinct
    user_ids = [r[0] for r in db.query(distinct(SupportMessage.user_id)).all()]
    result = []
    for uid in user_ids:
        user = db.get(User, uid)
        if not user:
            continue
        last = db.query(SupportMessage).filter(SupportMessage.user_id == uid).order_by(SupportMessage.created_at.desc()).first()
        unread = db.query(func.count(SupportMessage.id)).filter(
            SupportMessage.user_id == uid,
            SupportMessage.is_from_user == True, SupportMessage.is_read == False,
        ).scalar() or 0
        preview = last.attachment_name or last.content[:50] if last else ""
        result.append(ConversationOut(
            user_id=uid, username=user.username,
            last_message=f"[附件] {preview}" if last and last.attachment_url and not last.content else preview,
            last_at=last.created_at if last else datetime.utcnow(),
            unread_count=unread,
        ))
    result.sort(key=lambda x: x.last_at, reverse=True)
    return result


def _to_out(msg: SupportMessage, db: Session) -> MessageOut:
    sender = db.get(User, msg.sender_id)
    return MessageOut(
        id=msg.id, user_id=msg.user_id, sender_id=msg.sender_id,
        sender_name=sender.username if sender else "?",
        content=msg.content, is_from_user=msg.is_from_user,
        is_read=msg.is_read, created_at=msg.created_at,
        attachment_url=msg.attachment_url,
        attachment_name=msg.attachment_name,
        attachment_type=msg.attachment_type,
    )

File: backend-go/handlers/support.go
package handlers

import (
	"net/http"
	"os"
	"path/filepath"
	"crypto/rand"
	"encoding/hex"
	"strings"
	"time"

	"simnexus-go/config"
	"simnexus-go/database"
	"simnexus-go/middleware"
	"simnexus-go/models"
	"simnexus-go/security"
	"simnexus-go/services"

	"github.com/gin-gonic/gin"
)

const maxSupportFileSize = 20 * 1024 * 1024

var supportImageTypes = map[string]bool{
	"image/jpeg": true, "image/png": true, "image/gif": true,
	"image/webp": true, "image/bmp": true,
}

func supportMsgOut(m *models.SupportMessage) gin.H {
	var sender models.User
	name := "?"
	if database.DB.First(&sender, m.SenderID).Error == nil {
		name = sender.Username
	}
	return gin.H{
		"id": m.ID, "user_id": m.UserID, "sender_id": m.SenderID,
		"sender_name": name, "content": m.Content, "is_from_user": m.IsFromUser,
		"is_read": m.IsRead, "created_at": m.CreatedAt,
		"attachment_url": m.AttachmentURL, "attachment_name": m.AttachmentName,
		"attachment_type": m.AttachmentType,
	}
}

// SupportUpload stores an uploaded file and returns its URL.
func SupportUpload(c *gin.Context) {
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "缺少文件"})
		return
	}
	if file.Size > maxSupportFileSize {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "文件大小超过 20MB 限制"})
		return
	}
	os.MkdirAll(config.C.UploadDir, 0o755)
	ext := strings.ToLower(filepath.Ext(file.Filename))
	buf := make([]byte, 16)
	rand.Read(buf)
	name := hex.EncodeToString(buf) + ext
	dst := filepath.Join(config.C.UploadDir, name)
	if err := c.SaveUploadedFile(file, dst); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "保存失败"})
		return
	}
	ct := file.Header.Get("Content-Type")
	attType := "file"
	if supportImageTypes[ct] {
		attType = "image"
	}
	c.JSON(http.StatusOK, gin.H{"url": "/api/support/files/" + name, "name": file.Filename, "type": attType})
}

// SupportServeFile serves an uploaded file.
func SupportServeFile(c *gin.Context) {
	filename := c.Param("filename")
	if strings.Contains(filename, "/") || strings.Contains(filename, "..") {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "非法文件名"})
		return
	}
	path := filepath.Join(config.C.UploadDir, filename)
	if _, err := os.Stat(path); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "文件不存在"})
		return
	}
	c.File(path)
}

type messageIn struct {
	Content        string `json:"content"`
	UserID         *uint  `json:"user_id"`
	AttachmentURL  string `json:"attachment_url"`
	AttachmentName string `json:"attachment_name"`
	AttachmentType string `json:"attachment_type"`
}

func strPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

// SupportSendMessage posts a support chat message.
func SupportSendMessage(c *gin.Context) {
	me := middleware.CurrentUser(c)
	var body messageIn
	c.ShouldBindJSON(&body)
	if strings.TrimSpace(body.Content) == "" && body.AttachmentURL == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "消息或附件不能同时为空"})
		return
	}
	staff := security.IsSupportStaff(me)
	var msg models.SupportMessage
	if staff {
		if body.UserID == nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "客服必须指定目标用户"})
			return
		}
		var target models.User
		if database.DB.First(&target, *body.UserID).Error != nil {
			c.JSON(http.StatusNotFound, gin.H{"detail": "用户不存在"})
			return
		}
		msg = models.SupportMessage{
			UserID: *body.UserID, SenderID: me.ID,
			Content: strings.TrimSpace(body.Content), IsFromUser: false,
		}
	} else {
		msg = models.SupportMessage{
			UserID: me.ID, SenderID: me.ID,
			Content: strings.TrimSpace(body.Content), IsFromUser: true,
		}
	}
	msg.AttachmentURL = strPtr(body.AttachmentURL)
	msg.AttachmentName = strPtr(body.AttachmentName)
	msg.AttachmentType = strPtr(body.AttachmentType)
	msg.CreatedAt = time.Now()
	database.DB.Create(&msg)

	preview := strings.TrimSpace(body.Content)
	if len(preview) > 40 {
		preview = preview[:40]
	}
	if preview == "" {
		if body.AttachmentName != "" {
			preview = "[" + body.AttachmentName + "]"
		} else {
			preview = "[附件]"
		}
	}
	if staff {
		services.Push("support_reply", "客服已回复您的咨询", preview, "user", body.UserID)
	} else {
		services.Push("support_msg", "用户咨询："+me.Username, preview, "support", nil)
	}
	c.JSON(http.StatusOK, supportMsgOut(&msg))
}

// SupportGetMessages returns a conversation's messages.
func SupportGetMessages(c *gin.Context) {
	me := middleware.CurrentUser(c)
	q := database.DB.Model(&models.SupportMessage{})
	if security.IsSupportStaff(me) {
		uid := c.Query("user_id")
		if uid == "" {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "需要指定 user_id"})
			return
		}
		q = q.Where("user_id = ?", uid)
	} else {
		q = q.Where("user_id = ?", me.ID)
	}
	if since := c.Query("since_id"); since != "" {
		q = q.Where("id > ?", since)
	}
	var msgs []models.SupportMessage
	q.Order("created_at asc").Find(&msgs)
	out := make([]gin.H, 0, len(msgs))
	for i := range msgs {
		out = append(out, supportMsgOut(&msgs[i]))
	}
	c.JSON(http.StatusOK, out)
}

// SupportMarkRead marks a conversation read.
func SupportMarkRead(c *gin.Context) {
	me := middleware.CurrentUser(c)
	if security.IsSupportStaff(me) {
		uid := c.Query("user_id")
		if uid == "" {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "需要 user_id"})
			return
		}
		database.DB.Model(&models.SupportMessage{}).
			Where("user_id = ? AND is_from_user = ? AND is_read = ?", uid, true, false).
			Update("is_read", true)
	} else {
		database.DB.Model(&models.SupportMessage{}).
			Where("user_id = ? AND is_from_user = ? AND is_read = ?", me.ID, false, false).
			Update("is_read", true)
	}
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

// SupportUnread returns unread count.
func SupportUnread(c *gin.Context) {
	me := middleware.CurrentUser(c)
	var count int64
	if security.IsSupportStaff(me) {
		database.DB.Model(&models.SupportMessage{}).
			Where("is_from_user = ? AND is_read = ?", true, false).Count(&count)
	} else {
		database.DB.Model(&models.SupportMessage{}).
			Where("user_id = ? AND is_from_user = ? AND is_read = ?", me.ID, false, false).Count(&count)
	}
	c.JSON(http.StatusOK, gin.H{"count": count})
}

// SupportConversations lists conversations (staff only).
func SupportConversations(c *gin.Context) {
	me := middleware.CurrentUser(c)
	if !security.IsSupportStaff(me) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无客服权限"})
		return
	}
	var userIDs []uint
	database.DB.Model(&models.SupportMessage{}).Distinct("user_id").Pluck("user_id", &userIDs)
	type conv struct {
		UserID      uint      `json:"user_id"`
		Username    string    `json:"username"`
		LastMessage string    `json:"last_message"`
		LastAt      time.Time `json:"last_at"`
		UnreadCount int64     `json:"unread_count"`
	}
	out := make([]conv, 0)
	for _, uid := range userIDs {
		var user models.User
		if database.DB.First(&user, uid).Error != nil {
			continue
		}
		var last models.SupportMessage
		database.DB.Where("user_id = ?", uid).Order("created_at desc").First(&last)
		var unread int64
		database.DB.Model(&models.SupportMessage{}).
			Where("user_id = ? AND is_from_user = ? AND is_read = ?", uid, true, false).Count(&unread)
		preview := last.Content
		if last.AttachmentName != nil && *last.AttachmentName != "" {
			preview = *last.AttachmentName
		}
		if len(preview) > 50 {
			preview = preview[:50]
		}
		if last.AttachmentURL != nil && *last.AttachmentURL != "" && last.Content == "" {
			preview = "[附件] " + preview
		}
		out = append(out, conv{
			UserID: uid, Username: user.Username, LastMessage: preview,
			LastAt: last.CreatedAt, UnreadCount: unread,
		})
	}
	// sort by last_at desc
	for i := 0; i < len(out); i++ {
		for j := i + 1; j < len(out); j++ {
			if out[j].LastAt.After(out[i].LastAt) {
				out[i], out[j] = out[j], out[i]
			}
		}
	}
	c.JSON(http.StatusOK, out)
}
